### PR TITLE
Ensure we validate the model target

### PR DESCRIPTION
**Pull request by SimonRichardson**
Instead of using the charm URL, which will be wrong for bases, instead
use only the metadata. The metadata will always be right!

So the following code changes the purpose of this validate code,
from catching a non-container on a container workload. We drop the
series check because we validate that check in other places.

## QA steps

```sh
$ juju bootstrap lxd test
$ juju deploy hello-juju
```

### CODE REVIEW

**Comment by hmlanigan** [comment]
Combined with #12881:

$ juju deploy ambassador
Located charm "ambassador" in charm-hub, revision 13
Deploying "ambassador" from charm-hub charm "ambassador", revision 13 in channel stable
$ juju deploy facundo-snappass-test
Located charm "facundo-snappass-test" in charm-hub, revision 3
Deploying "facundo-snappass-test" from charm-hub charm "facundo-snappass-test", revision 3 in channel stable

Still need to figure out the kubernetes bundles

File: core/model/model.go [hunk 1]
```diff
@@ -41,32 +39,27 @@ type Model struct {
...
-			return errors.Trace(err)
-		}
-	} else {
-		os, err := series.GetOSFromSeries(charmSeries)
-		if err != nil {
-			return errors.Trace(err)
+	switch modelType {
+	case CAAS:
+		if isIAAS {
+			return errors.NotValidf("non container base charm for container based model type")
```
Comment (manadart) [inline core/model/model.go:55]: "container-based"

File: core/model/model.go [hunk 2]
```diff
@@ -41,32 +39,27 @@ type Model struct {
...
-			if !caasOS.Contains(os.String()) {
-				return errors.NotValidf("series %q in a kubernetes model", charmSeries)
-			}
-		case IAAS:
-			if caasOS.Contains(os.String()) {
-				return errors.NotValidf("series %q in a non container model", charmSeries)
-			}
+	case IAAS:
+		if !isIAAS {
+			return errors.NotValidf("container base charm for non container based model type")
```
Comment (manadart) [inline core/model/model.go:59]: And here.

**Review by manadart — APPROVED** [review]
Much better.

We're looking sweet now.

On LXD:
```
$ juju deploy hello-juju                       
located charm "hello-juju" in charm-hub, revision 6
deploying "hello-juju" from charm-hub charm "hello-juju", revision 6 in channel stable
$ juju deploy ambassador
located charm "ambassador" in charm-hub, revision 13
deploying "ambassador" from charm-hub charm "ambassador", revision 13 in channel stable
error cannot add application "ambassador": container base charm for non container based model type not valid
$ juju deploy facundo-snappass-test
located charm "facundo-snappass-test" in charm-hub, revision 3
deploying "facundo-snappass-test" from charm-hub charm "facundo-snappass-test", revision 3 in channel stable
error cannot add application "facundo-snappass-test": container base charm for non container based model type not valid
```

On Microk8s:
```
$ juju deploy facundo-snappass-test
Located charm "facundo-snappass-test" in charm-hub, revision 3
Deploying "facundo-snappass-test" from charm-hub charm "facundo-snappass-test", revision 3 in channel stable
$ juju deploy ambassador
Located charm "ambassador" in charm-hub, revision 13
Deploying "ambassador" from charm-hub charm "ambassador", revision 13 in channel stable
$ juju deploy hello-juju
Located charm "hello-juju" in charm-hub, revision 6
Deploying "hello-juju" from charm-hub charm "hello-juju", revision 6 in channel stable
ERROR cannot add application "hello-juju": non container base charm for container based model type not valid
```

Everything quiesced as expected.

File: state/export_test.go [hunk 3]
```diff
@@ -272,9 +272,19 @@ func addTestingApplication(c *gc.C, params addTestingApplicationParams) *Applica
 	return app
 }
 
-func addCustomCharm(c *gc.C, st *State, repo *charmrepotesting.Repo, name, filename, content, series string, revision int) *Charm {
+func addCustomCharmWithManifest(c *gc.C, st *State, repo *charmrepotesting.Repo, name, filename, content, series string, revision int, manifest bool) *Charm {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply state/export_test.go:275]: I agree, but I don't think we have much option as this stuff is just getting silly now. The code in question tests that the charm loaded is kubernetes, but then uses a totally different metadata. I'm kinda thinking that most of this code should be nuked from up high.

**Pull request merged**